process_request/routines.py

```diff
@@ -29,36 +32,46 @@ def get_data(self, item):
                         username=settings.ARCHIVESSPACE["username"],
                         password=settings.ARCHIVESSPACE["password"],
                         repository=settings.ARCHIVESSPACE["repo_id"])
-        obj = aspace.client.get(item)
+        obj = aspace.client.get(item, params={"resolve": ["resource::linked_agents", "ancestors",
+                                                          "top_container", "top_container::container_locations", "instances::digital_object"]})
+
         if obj.status_code == 200:
-            return obj.json()
+            as_data = {}
+            item_json = obj.json()
+            item_collection = item_json.get("ancestors")[-1].get("_resolved")
+            as_data['creator'] = get_collection_creator(item_collection)
+            as_data['restrictions'] = "TK"
+            as_data['restrictions_text'] = "TK"
+            as_data['collection_name'] = item_collection.get("title")
+            if len(item_json.get("ancestors")) > 1:
+                as_data['aggregation'] = item_json.get("ancestors")[0].get("_resolved").get("display_string")
+            else:
+                as_data['aggregation'] = ""
+            as_data['dates'] = get_dates(item_json)
+            as_data['resource_id'] = item_collection.get("id_0")
+            as_data['title'] = item_json.get("display_string")
+            as_data['ref'] = item_json.get("uri")
+            instances = item_json.get("instances")
+            if instances:
+                container_indicators = []
+                for i in instances:
+                    [container_indicators.append(get_container_indicators(i))]
+                as_data['containers'] = ", ".join(container_indicators)
+                preferred_item = get_preferred_format(instances)
+                if preferred_item:
+                    set_preferred_data(as_data,
+                                       preferred_item[0]['indicator'],
+                                       preferred_item[0]['instance_type'],
+                                       preferred_item[0]['location'])
+                else:
+                    set_preferred_data(as_data)
+            else:
+                as_data['containers'] = None
+                set_preferred_data(as_data)
```

Comment:
This logic is super complex and I'm not sure I understand why?

Reply:
So, I think most of the complexity comes from the assumption that if I'm working with an `as_data` dict, that we'd want certain keys to show up even if they're empty.

First, I'm getting a list of all indicators joined together.

Then, I'm using the instances info to grab the preferred format. It's separated out from the `for i in instances` loop because I'm getting a list of `instance_types` in that method. However, I could just as easily remove that from the method and have `get_preferred_format` to use a single instance instead.

Then, the ending is kinda complex because of the assumption that we'd want those keys. As it has been written, there will not be a `preferred_item` in the cases where an instance type was not digital object, microform, or mixed materials. So, it was going to error if you based it av materials or something else. We could remove that check if we're returning something no matter what. That would also eliminate the need for the else statement.

The else statement at the end also assumes we'd want those keys even if they're set to `None`. They weren't showing up in `as_data` unless specifically set.

Reply:
OK, so first I'd think about encapsulating all the logic you need in your helper methods so that you always get the desired value back. That way you can keep this `get_data` method fairly clean, and it will be easier to read and debug the code in the future. So, for example, your `get_container_indicators` method should just return None if there are no instances, same for `get_preferred_format`.

Regarding your assumptions, I think it's true that we'd want keys to exist even if their value is null. However, I think that we *do* always want to return preferred instance values. If there aren't digital or microfilm instances then it's just whatever else is there.